### PR TITLE
Optimize balancer and setup debug logger.

### PR DESCRIPTION
**Pull request by JacoCheung**
## Description
This PR optimizes the HSTU batch balancer pipeline to fully hide the KK (Karmarkar-Karp) algorithm overhead and reduce AllGather communication cost.

### Core Changes

**Async 2-phase KK pipeline**
- Split `shuffle()` into `start_shuffle_async()` (Phase 1) + `finish_shuffle()` (Phase 2).
- Phase 1 runs AllGather workloads (NCCL) on main thread, then submits KK partitioning to a background `ThreadPoolExecutor` (pure CPU, no GPU access).
- Phase 2 waits for KK result, then redistributes batch data (AllGather or All2All).
- KK computation fully overlaps with forward/backward on the GPU.
- `ShuffleHandle` provides stable per-batch tracking for the prefetch pipeline (multiple in-flight batches).

**All2All shuffle path** (`batch_all2all.py`)
- New O(B) communication alternative to AllGather+index_select.
- Toggled via `SHUFFLE_WITH_ALL2ALL=1` env var.
- Local `dst_rank` derivation (no extra communication when partitions are symmetric).

**Fused multi-KJT AllGather** (`collective_ops.py`)
- `keyed_jagged_tensor_list_allgather`: gathers all KJTs with only 2 NCCL calls (1 for lengths, 1 for values), regardless of KJT count.
- Batch size and dtype validation added for input KJTs.

**Pipeline integration** (`train_pipeline.py`)
- Both `JaggedMegatronTrainPipelineSparseDist` and `JaggedMegatronPrefetchTrainPipelineSparseDist` updated with the 2-phase async API.
- NCCL ordering guards (`wait_stream`) prevent deadlocks between shuffle and DDP/loss all-reduces.

### Performance & Debug Utilities

- **Attention perf tracker** (`attn_perf_tracker.py`, `hooks.py`): CUDA-event-based per-layer TFLOPS/MFU logging via module hooks (`PRINT_HSTU_PERF=1`).
- **GPU device spec & MFU** (`perf.py`): `DeviceSpec` auto-discovery, peak TFLOPS calculation, MFU reporting.
- **Balancer logger**: load-balance before/after stats, controlled by `PRINT_LOAD_BALANCE=1` with start/stop window.
- **Dataset distribution support**: `UNIFORM`, `NORMAL`, `ZIPF`, `LOGNORMAL` seqlen distributions for `HSTURandom*Dataset`.
- **Balanced attention benchmark**: standalone benchmark for HSTU kernel SoL and balancer speedup analysis.

### Bug Fixes (from review)

- Removed accidental `pdb.set_trace()` left in `test_dataset.py`.
- Fixed FLOP double-counting: removed inline perf tracking from `FusedHSTULayerFunction`, keeping only module-level hooks.
- Converted `_batch_counter` from class variable to instance variable.
- Fixed off-by-one in async path print window check (pass handle idx explicitly).
- Removed unused `linear_dim` parameter from `_compute_attn_fwd_flops`.
- Added `actual_batch_size` documentation note in `pad_and_allgather_batch`.

---

## CI

- [Pipeline #45786214](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/45786214) — 2026-03-10 10:26 UTC
- [Pipeline #45865127](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/45865127) — 2026-03-11 08:17 UTC
- [Pipeline #45940124](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/45940124) — 2026-03-12 03:34 UTC
- [Pipeline #45957243](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/45957243) — 2026-03-12 08:14 UTC
- [Pipeline #46023804](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/46023804) — 2026-03-13 02:14 UTC

### CODE REVIEW

File: examples/hstu/training/benchmark/run_balanced_pipeline_comparison.sh [hunk 1]
```diff
@@ -0,0 +1,225 @@
+#!/bin/bash
```
Comment (shijieliu) [inline examples/hstu/training/benchmark/run_balanced_pipeline_comparison.sh]: do we need to update benchmark doc?

Comment (shijieliu) [reply examples/hstu/training/benchmark/run_balanced_pipeline_comparison.sh]: and I’m a little worried that these benchmark scripts will soon break if we don’t add them to CI or put some other monitoring in place. But if we add it in CI I feel it's too heavy. Any idea?

Comment (JacoCheung) [reply examples/hstu/training/benchmark/run_balanced_pipeline_comparison.sh]: Removed

File: examples/commons/distributed/batch_shuffler.py [hunk 2]
```diff
@@ -39,6 +91,20 @@ def shuffle(
         partitions_indices = karmarkar_karp(
             allgather_workloads, num_partitions, equal_size=True
         )
+        if self._should_print_load_balance():
+            all_wl = (
+                allgather_workloads.tolist()
+                if isinstance(allgather_workloads, torch.Tensor)
+                else list[Any](allgather_workloads)
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py]: Not applicable — our minimum Python is 3.10+, `list[Any]` is valid. The import line was changed to include `Optional` for other reasons.

File: examples/commons/utils/logger.py [hunk 3]
```diff
@@ -28,13 +29,57 @@
...
+    """If distributed is initialized, print only on rank 0."""
+    if torch.distributed.is_initialized():
+        if torch.distributed.get_rank() == 0:
+            _logger.debug(message)
+    else:
+        print(message, flush=True)
+
+
+def print_rank_all(message, level=logging.INFO):
+    """If distributed is initialized, print only on rank 0."""
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/utils/logger.py]: Fixed in earlier commit — docstrings now say "print on all ranks".

File: examples/commons/utils/logger.py [hunk 4]
```diff
@@ -28,13 +29,57 @@
...
+    if torch.distributed.is_initialized():
+        _logger.log(level, message)
+    else:
+        print(message, flush=True)
+
+
+def info_rank_all(message):
+    """If distributed is initialized, print only on rank 0."""
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/utils/logger.py]: Fixed.

File: examples/commons/utils/logger.py [hunk 5]
```diff
@@ -28,13 +29,57 @@
...
+    if torch.distributed.is_initialized():
+        _logger.info(message)
+    else:
+        print(message, flush=True)
+
+
+def debug_rank_all(message):
+    """If distributed is initialized, print only on rank 0."""
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/utils/logger.py]: Fixed.

File: examples/commons/distributed/batch_shuffler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py]: Not applicable — our minimum Python is 3.10+, `list[Any]` is valid.

File: examples/commons/utils/logger.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/utils/logger.py]: Fixed in earlier commit — docstrings now say "print on all ranks".

File: examples/commons/utils/logger.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/utils/logger.py]: Fixed.

File: examples/commons/utils/logger.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/utils/logger.py]: Fixed.

File: examples/commons/pipeline/train_pipeline.py [hunk 6]
```diff
@@ -712,6 +712,28 @@ def __init__(
...
             batch_shuffler=batch_shuffler,
         )
+        # Background thread for H2D + batch shuffle so it doesn't block
+        # the main thread's forward / backward.
+        import concurrent.futures
+
+        self._shuffle_executor = concurrent.futures.ThreadPoolExecutor(
+            max_workers=1, thread_name_prefix="shuffle"
+        )
+        self._shuffle_future: Optional["concurrent.futures.Future[Optional[In]]"] = None
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/pipeline/train_pipeline.py]: Already handled — `BaseTaskBalancedBatchShuffler.__del__` calls `self._kk_executor.shutdown(wait=False)`. The executor is in the shuffler, not the pipeline.

File: examples/commons/distributed/batch_shuffler.py [hunk 7]
```diff
@@ -1,36 +1,260 @@
...
+            aw = meta["allgather_workloads"]
+            all_wl = aw.tolist() if isinstance(aw, torch.Tensor) else list(aw)
+            _log_load_balance(
+                self._batch_counter,
+                all_wl,
+                partitions_indices,
+                meta["local_batch_size"],
+                meta["num_partitions"],
+            )
+        self._batch_counter += 1
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py]: Fixed — `finish_shuffle` no longer increments `_batch_counter`. The counter is incremented only in `start_shuffle_async` (for the async path) or `compute_partition_indices` (for the sync path).

File: examples/commons/distributed/batch_shuffler.py [hunk 8]
```diff
@@ -1,36 +1,303 @@
...
+        f"[Load Balance] batch={batch_idx}\n"
+        f"Before: all_ranks=[{', '.join(f'{x:.3e}' for x in before_loads)}]  "
+        f"max-min={before_max - before_min:.3e}  imbalance={before_imb:.2f}%\n"
+        f"After:  all_ranks=[{', '.join(f'{x:.3e}' for x in after_loads)}]  "
+        f"max-min={after_max - after_min:.3e}  imbalance={after_imb:.2f}%",
+    )
+
 
 class BaseTaskBalancedBatchShuffler:
+    _batch_counter: int = 0
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py]: Fixed in 5fdc671 — `_batch_counter` is now an instance variable initialized in `__init__`.

File: examples/commons/distributed/batch_shuffler.py [hunk 9]
```diff
@@ -1,36 +1,315 @@
...
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py]: Fixed in 5fdc671 — converted to instance variable.

File: examples/commons/ops/collective_ops.py [hunk 10]
```diff
@@ -501,13 +504,143 @@ def jagged_tensor_allgather(
...
+        return []
+
+    W = world_size
+    device = kjt_list[0].lengths().device
+
+    # --- Collect per-KJT metadata ---
+    keys_list = [list(kjt.keys()) for kjt in kjt_list]
+    K_list = [len(keys) for keys in keys_list]
+    K_total = sum(K_list)
+    B = kjt_list[0].lengths().numel() // K_list[0]
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/ops/collective_ops.py:549]: Fixed in 5fdc671 — added batch_size and values dtype validation assertions before concatenation.

File: examples/commons/ops/collective_ops.py [hunk 11]
```diff
@@ -501,13 +504,143 @@ def jagged_tensor_allgather(
...
+
+    # --- Concatenate all lengths and values across KJTs ---
+    all_local_lengths = torch.cat([kjt.lengths() for kjt in kjt_list])  # [K_total * B]
+    all_local_values = torch.cat([kjt.values() for kjt in kjt_list])  # [T_total_local]
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/ops/collective_ops.py:564]: Fixed in 5fdc671 — added dtype validation.

File: examples/commons/distributed/batch_shuffler.py [hunk 12]
```diff
@@ -1,36 +1,303 @@
...
+        # 2. Python object IDs can be reused, leading to collisions
+        # 3. ShuffleHandle provides a stable, monotonically increasing identifier
+        self._kk_states: Dict[ShuffleHandle, Dict[str, Any]] = {}
+
+    def __del__(self) -> None:
+        """Clean up ThreadPoolExecutor on object destruction."""
+        if hasattr(self, "_kk_executor"):
+            # Shutdown executor, wait=False to avoid blocking during cleanup
+            # The executor will finish current tasks but won't accept new ones
+            self._kk_executor.shutdown(wait=False)
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py:161]: Intentional — `wait=False` is used in `__del__` to avoid blocking Python's garbage collection. The executor has `max_workers=1` and only runs pure-CPU KK tasks, so a straggling task will complete quickly and is harmless.

File: examples/commons/distributed/batch_shuffler.py [hunk 13]
```diff
@@ -1,36 +1,315 @@
...
+    def _should_print_load_balance(self) -> bool:
+        """Check if load balance info should be printed for the current batch."""
+        if not _PRINT_LOAD_BALANCE:
+            return False
+        idx = self._batch_counter
+        if idx < _PRINT_LOAD_BALANCE_START:
+            return False
+        if _PRINT_LOAD_BALANCE_STOP >= 0 and idx >= _PRINT_LOAD_BALANCE_STOP:
+            return False
+        return True
```
[the remark this answers is not in the document]
Comment (JacoCheung) [reply examples/commons/distributed/batch_shuffler.py]: Fixed in 5fdc671 — `_should_print_load_balance` now accepts an optional `idx` parameter. The async path (`finish_shuffle`) passes `int(handle)` explicitly instead of relying on the eagerly-incremented `_batch_counter`.
